core/nightshadev2/src/nightshade_task.rs

```diff
@@ -0,0 +1,194 @@
+use std::cmp::min;
+use std::collections::HashSet;
+use std::time::Duration;
+use std::time::Instant;
+
+use futures::Async;
+use futures::future::Future;
+use futures::Poll;
+use futures::try_ready;
+use futures::sink::Sink;
+use futures::Stream;
+use futures::sync::mpsc;
+use log::{debug, error, info};
+use tokio::timer::Delay;
+
+use super::nightshade::{AuthorityId, Nightshade, State};
+
+pub enum Control {
+    Reset,
+    Stop,
+}
+
+#[derive(Clone)]
+pub struct Message {
+    pub author: AuthorityId,
+    pub receiver_id: AuthorityId,
+    pub state: State,
+}
+
+const COOLDOWN_MS: u64 = 50;
+const GOSSIP_SAMPLE_SIZE: usize = 3;
+
+pub struct NightshadeTask {
+    owner_id: AuthorityId,
+    num_authorities: usize,
+    nightshade: Option<Nightshade>,
+    state_receiver: mpsc::Receiver<Message>,
+    state_sender: mpsc::Sender<Message>,
+    control_receiver: mpsc::Receiver<Control>,
+
+    /// Timer that determines the minimum time that we should not gossip after the given message
+    /// for the sake of not spamming the network with small packages.
+    cooldown_delay: Option<Delay>,
+}
+
+impl NightshadeTask {
+    pub fn new(
+        owner_id: AuthorityId,
+        num_authorities: usize,
+        state_receiver: mpsc::Receiver<Message>,
+        state_sender: mpsc::Sender<Message>,
+        control_receiver: mpsc::Receiver<Control>,
+    ) -> Self {
+        Self {
+            owner_id,
+            num_authorities,
+            nightshade: None,
+            state_receiver,
+            state_sender,
+            control_receiver,
+            cooldown_delay: None,
+        }
+    }
+
+    fn nightshade_as_ref(&self) -> &Nightshade {
+        self.nightshade.as_ref().expect("Nightshade should be initialized")
+    }
+
+    fn nightshade_as_mut_ref(&mut self) -> &mut Nightshade {
+        self.nightshade.as_mut().expect("Nightshade should be initialized")
+    }
+
+    fn init_nightshade(&mut self) {
+        self.nightshade = Some(
+            Nightshade::new(self.owner_id as AuthorityId, self.num_authorities as usize));
+    }
+
+    fn state(&self) -> State {
+        self.nightshade_as_ref().state()
+    }
+
+    fn send_state(&self, message: Message) {
+        let copied_tx = self.state_sender.clone();
+        tokio::spawn(copied_tx.send(message).map(|_| ()).map_err(|e| {
+            error!("Error sending state. {:?}", e);
+        }));
+    }
+
+    #[allow(dead_code)]
+    fn is_final(&self) -> bool {
+        self.nightshade_as_ref().is_final()
+    }
+
+    fn process_message(&mut self, message: Message) {
+        self.nightshade_as_mut_ref().update_state(message.author, message.state);
+    }
+
+    /// Sends gossip to random authority peers.
+    fn gossip_state(&self) {
```

Comment:
send message to everyone for now - in this version we only send our own state and everyone needs to see it

Reply:
Done [here](https://github.com/nearprotocol/nearcore/pull/549/commits/4dda84d65ad03ae36c8fa45761bb97425c7b6560)